File: lua/distract/manifests/sun.lua
-- Frame indices come from the generated sprite layout rather than being
-- written out by hand, so the manifest and the art cannot drift apart.
local layout = require("distract.terminal_sprites").get_layout("sun")

local M = {
  name = "sun",
  asset_type = "procedural",
  spritesheet = {},
  initial_state = "shining",
  locomotion = "omnidirectional",
  capabilities = { locomotion = { "omnidirectional" } },
  -- A sun belongs in the sky, and that is a fact about suns rather than about
  -- anyone's configuration. `omnidirectional` only says it is free to leave the
  -- floor, which on its own put it in the middle of the screen. A spawn or a
  -- `position.anchor` still overrides this.
  anchor = "top",
  z_index = -10,
  states = {
    shining = {
      animation = { frames = layout.shining, fps = 2.0, loop_anim = true, flip_x = false },
      physics = { target_vx = 0.2, target_vy = 0.0, wrap_mode = "wrap", path_type = "sine" },
      transitions = {
        on_event = {
          scrolling = "flare",
        },
      },
    },
    rising = {
      -- One-shot transition: it plays through and holds, then times out to
      -- shining. Looping it would snap the sun back to its start pose.
      animation = { frames = layout.rising, fps = 2.0, loop_anim = false, flip_x = false },
      physics = { target_vx = 0.5, target_vy = -1.0, wrap_mode = "clamp" },
      transitions = {
        timeout_ms = 4000,
        on_timeout = "shining",
      },
    },
    setting = {
      -- One-shot transition: it plays through and holds, then times out to
      -- shining. Looping it would snap the sun back to its start pose.
      animation = { frames = layout.setting, fps = 2.0, loop_anim = false, flip_x = false },
      physics = { target_vx = 0.5, target_vy = 1.0, wrap_mode = "clamp" },
      transitions = {
        timeout_ms = 4000,
        on_timeout = "shining",
      },
    },
    eclipse = {
      -- One-shot transition: it plays through and holds, then times out to
      -- shining. Looping it would snap the sun back to its start pose.
      animation = { frames = layout.eclipse, fps = 1.0, loop_anim = false, flip_x = false },
      physics = { target_vx = 0.0, target_vy = 0.0 },
      transitions = {
        timeout_ms = 8000,
        on_timeout = "shining",
      },
    },
    flare = {
      animation = { frames = layout.flare, fps = 6.0, loop_anim = false, flip_x = false },
      physics = { target_vx = 0.4, target_vy = 0.0 },
      transitions = {
        on_finish = "shining",
        timeout_ms = 2000,
        on_timeout = "shining",
      },
    },
  },
  custom_actions = {
    eclipse = {
      target_state = "eclipse",
      duration_ms = 8000,
      return_state = "shining",
    },
    rise = {
      target_state = "rising",
      duration_ms = 4000,
      return_state = "shining",
    },
    set = {
      target_state = "setting",
      duration_ms = 4000,
      return_state = "shining",
    },
    flare = {
      target_state = "flare",
      duration_ms = 2000,
      return_state = "shining",
    },
  },
}

return M

File: engine/src/manifests/sun.rs
//! The built-in sun's manifest.
//!
//! A static state table and nothing else. §5 of the standards exempts pure data
//! tables in dedicated files from the size cap, and keeping all three of these in
//! `manifest.rs` is what put that module past 1,400 lines.

use std::collections::HashMap;

use crate::manifest::{
    AnimationConfig, AssetManifest, Capabilities, CustomActionConfig, OMNIDIRECTIONAL,
    PhysicsConfig, SpritesheetConfig, StateDefinition, TransitionConfig, WrapMode,
};
use crate::sprites;

/// Built-in Sun manifest with shining, rise, set, eclipse, and solar flare.
pub fn manifest() -> AssetManifest {
    let mut states = HashMap::new();

    // Shining (default state)
    let mut shine_trans = HashMap::new();
    shine_trans.insert("scrolling".to_string(), "flare".to_string());

    states.insert(
        "shining".to_string(),
        StateDefinition {
            animation: AnimationConfig {
                frames: sprites::sun_set().frames_for("shining"),
                fps: 2.0,
                loop_anim: true,
                flip_x: false,
            },
            physics: PhysicsConfig {
                target_vx: 0.2,
                target_vy: 0.0,
                wrap_mode: WrapMode::Wrap,
                path_type: Some("sine".to_string()),
                path_amplitude: Some(15.0),
                path_frequency: Some(2.0),
                ..Default::default()
            },
            transitions: TransitionConfig {
                on_event: shine_trans,
                ..Default::default()
            },
            is_locked: false,
        },
    );

    // Rising (moves upwards)
    states.insert(
        "rising".to_string(),
        StateDefinition {
            animation: AnimationConfig {
                frames: sprites::sun_set().frames_for("rising"),
                fps: 2.0,
                loop_anim: true,
                flip_x: false,
            },
            physics: PhysicsConfig {
                target_vx: 0.5,
                target_vy: -1.0,
                wrap_mode: WrapMode::Clamp,
                ..Default::default()
            },
            transitions: TransitionConfig {
                timeout_ms: Some(4000),
                on_timeout: Some("shining".to_string()),
                ..Default::default()
            },
            is_locked: true,
        },
    );

    // Setting (moves downwards)
    states.insert(
        "setting".to_string(),
        StateDefinition {
            animation: AnimationConfig {
                frames: sprites::sun_set().frames_for("setting"),
                fps: 2.0,
                loop_anim: true,
                flip_x: false,
            },
            physics: PhysicsConfig {
                target_vx: 0.5,
                target_vy: 1.0,
                wrap_mode: WrapMode::Clamp,
                ..Default::default()
            },
            transitions: TransitionConfig {
                timeout_ms: Some(4000),
                on_timeout: Some("shining".to_string()),
                ..Default::default()
            },
            is_locked: true,
        },
    );

    // Eclipse (moon blocks sun corona)
    states.insert(
        "eclipse".to_string(),
        StateDefinition {
            animation: AnimationConfig {
                frames: sprites::sun_set().frames_for("eclipse"),
                fps: 1.0,
                loop_anim: true,
                flip_x: false,
            },
            physics: PhysicsConfig {
                target_vx: 0.0,
                target_vy: 0.0,
                ..Default::default()
            },
            transitions: TransitionConfig {
                timeout_ms: Some(8000),
                on_timeout: Some("shining".to_string()),
                ..Default::default()
            },
            is_locked: true,
        },
    );

    // Solar flare
    states.insert(
        "flare".to_string(),
        StateDefinition {
            animation: AnimationConfig {
                frames: sprites::sun_set().frames_for("flare"),
                fps: 6.0,
                loop_anim: false,
                flip_x: false,
            },
            physics: PhysicsConfig {
                target_vx: 0.4,
                target_vy: 0.0,
                ..Default::default()
            },
            transitions: TransitionConfig {
                on_finish: Some("shining".to_string()),
                timeout_ms: Some(2000),
                on_timeout: Some("shining".to_string()),
                ..Default::default()
            },
            is_locked: true,
        },
    );

    let mut custom_actions = HashMap::new();
    custom_actions.insert(
        "eclipse".to_string(),
        CustomActionConfig {
            target_state: "eclipse".to_string(),
            duration_ms: Some(8000),
            return_state: Some("shining".to_string()),
            is_locked: Some(true),
        },
    );
    custom_actions.insert(
        "rise".to_string(),
        CustomActionConfig {
            target_state: "rising".to_string(),
            duration_ms: Some(4000),
            return_state: Some("shining".to_string()),
            is_locked: Some(true),
        },
    );
    custom_actions.insert(
        "set".to_string(),
        CustomActionConfig {
            target_state: "setting".to_string(),
            duration_ms: Some(4000),
            return_state: Some("shining".to_string()),
            is_locked: Some(true),
        },
    );
    custom_actions.insert(
        "flare".to_string(),
        CustomActionConfig {
            target_state: "flare".to_string(),
            duration_ms: Some(2000),
            return_state: Some("shining".to_string()),
            is_locked: Some(true),
        },
    );

    AssetManifest {
        name: "sun".to_string(),
        asset_type: "procedural".to_string(),
        spritesheet: SpritesheetConfig::default(),
        initial_state: "shining".to_string(),
        states,
        custom_actions,
        z_index: Some(-10),
        capabilities: Capabilities {
            locomotion: Some(vec![OMNIDIRECTIONAL.to_string()]),
        },
        locomotion: Some(OMNIDIRECTIONAL.to_string()),
        render: None,
    }
}
